packages/vite/src/node/optimizer/index.ts

```diff
@@ -1350,3 +1352,25 @@ const safeRename = promisify(function gracefulRename(
     if (cb) cb(er)
   })
 })
+
+export function tryOptimizedDepResolve(
+  config: ResolvedConfig,
+  ssr: boolean,
+  url: string,
+  depId: string,
+  fsResolveOptions: InternalResolveOptions,
+): string | undefined {
+  const depsOptimizer = getDepsOptimizer(config, ssr)
+
+  if (depsOptimizer?.isOptimizedDepFile(depId)) {
+    const depFile = cleanUrl(depId)
+    const info = optimizedDepInfoFromFile(depsOptimizer.metadata, depFile)
+    const depSrc = info?.src
+    if (depSrc) {
+      const resolvedFile = path.resolve(path.dirname(depSrc), url)
```

Comment:
Sorry it took me a while to review this.

IIUC this code assumes the optimized file and the original file to be a 1-to-1 mapping. But that isn't true. For example, if a dependency has multiple files, it becomes a single file.

When multiple file gets bundled in a single file, we need sourcemaps (or something that includes that kind of information) to identify which file the line was defined in.

I can think of 3 ways to fix this:

1. Load the sourcemap in [`optimizedDepsPlugin`'s load hook](https://github.com/vitejs/vite/blob/14e14726d11ac55c37cc2d15aaf935be18ef925b/packages/vite/src/node/plugins/optimizedDeps.ts#L71) and call `getCombinedSourcemap` in `assetImportMetaUrlPlugin` and use that to map to the original file
   -  the downside of this is that when other plugins modify the sourcemap incorrectly, the generated code will be incorrect
1. Save a modified optimized file when optimizing so that `assetImportMetaUrlPlugin` can resolve it correctly (like I tried to do in #13501)
   - the downside of this is that the optimize logic becomes more complicated
1. Modify the optimized file in `optimizedDepsPlugin`'s load hook so that `assetImportMetaUrlPlugin` can resolve it correctly
    - the downside of this is that there will be a overhead of running the transformation every time the server starts

Reply:
> IIUC this code assumes the optimized file and the original file to be a 1-to-1 mapping. But that isn't true. For example, if a dependency has multiple files, it becomes a single file.

I was thinking this a bit through and that's indeed tricky. Can you assist spinning up a failing test showing your exact concern? I'm not so deep into how splitting/merging might happen into Vite and keep mixing up things in my head on where we'd need to handle things differently. 

My thoughts: 

* According to the [rollup docs](https://rollupjs.org/plugin-development/#transform) the ids we get into the transform hook are referring to a single module which is transformed. 
* In `dev` mode we seem to always get URLs pointing to individual dependency files e.g. in a react app referencing my library I get something like `D:/Dev/myproject/src/node_modules/.vite/deps/@coderline_alphatab.js?v=9f8947bd`
   * These IDs originate from the dependencies loaded [before bundling](https://github.com/vitejs/vite/blob/012490ca8682e2b560737cb54dbb465ab4f36471/packages/vite/src/node/optimizer/index.ts#L391).
       * `addOptimizedDepInfo` is called with these IDs from via this [section](https://github.com/vitejs/vite/blob/012490ca8682e2b560737cb54dbb465ab4f36471/packages/vite/src/node/optimizer/optimizer.ts#L234-L250)
   * In dev-time the invidiual files of the dependencies are seemingly not bundled yet but served individually via `.vite/deps/` urls. 
* In `build` mode I get the real file path `D:/Dev/myproject/src/vite-react/node_modules/@coderline/alphatab/dist/alphaTab.mjs`
   * Looking at the output there is afterwards only a "index.js" bundle with all 3rd party modules and the app code bundled. 
* I generally assume any bundling/splitting should already be handled by Vite and Rollup correctly as things should be correct for any main application files. 
  * For main application files:
      * If files are not bundled, the modules are resolved via `tryFsResolve` correctly and from there we can find the worker/asset with the relative paths. 
      * If files are bundled/chunked, Vite needs to generate certain files (optimized deps area?). But I'm not sure how in this case the existing code would already work. 
  * For 3rd party files: 
      * If they are not bundled, we newly as a fallback try go back into node_modules and try to resolve the asset/worker relatively to there. This is what this PR is trying to achieve. 
      * If they are bundled/chunked we'd need to handle things more complex. Similarly to the main application file bundling I'm not sure how this would look like. 

My questions: 

* Trying to improve things step-by-step: Can we somehow detect the 1:1 and allow this path.  and 1:n/n:1 cases and for now show a warning of the 1:n/n:1 cases not being supported. 
* How would we reach the 1:n/n:1 path? In none of the frameworks or tests I was doing (dev/build) I managed to hit a path where bundling/chunking happened before this transform path was executed. 
    * For the main file importing the worker/asset: Is there a stage where the 3rd party modules are bundled/split before being transformed and where is the related information available? 
    * Is the `optimizedDepInfoFromFile` misleading and we need to loop through all matches as 'candidates' for valid paths? (not just pick the first one) 
* The existing code also mainly handles 1:1 scenarios, I wonder where the 1:n/n:1 logic is placed (other plugins doing additional rewrites?).

Reply:
> Can you assist spinning up a failing test showing your exact concern?

Sure. I pushed a commit that updates the test that shows my concern.

> I'm not so deep into how splitting/merging might happen into Vite ...

To clarify, the splitting/merging happens by the optimizer. Vite uses esbuild's bundle feature to bundle some dependencies and pass the bundled dependencies to the plugin pipeline (the plugin pipeline is similar to rollup). `new URL(, import.meta.url)` becomes invalid after the optimization (esbuild's bundling) because the optimizer outputs the dependency into a different directory and also merges the files.

Reply:
If we go with 2 or 3 in https://github.com/vitejs/vite/pull/16418#discussion_r1661124403, then we don't need this PR's change because the plugins won't need to change the way to resolve `new URL(, import.meta.url)`.  If we go with 1 in https://github.com/vitejs/vite/pull/16418#discussion_r1661124403, then we'll need a change like this PR but it won't use `OptimizedDepInfo.src` and I imagine the code will be quite different.

Reply:
This depends on how the transformed/corrected URL would look like whether this PR is needed or not. See this comparison: 

**Before Bundling**

_playground/assets/dep-with-asset/index.js_
```js
export { imageUrl } from './dir/path.js'
export const imageUrlRoot = new URL('./asset.png', import.meta.url).href;
```

_playground/assets/dep-with-asset/dir/path.js_
```js
export const imageUrl = new URL('../asset.png, import.meta.url).href;
```

**After Bundling (my current thinking)**

The URLs are adjusted to stay relative urls within the dependency just as-if the package would deliver the bundle itself. 

Pro: 
* This way we follow more a "relative **URL**" instead of a "relative **file path**" approach. 
* The bundler doesn't work in a way to serve the specific plugins for assets and workers. It is the concern of the asset & worker plugins that the URLs point to a physical file in the package and we need to adjust the paths so that the physical files are copied and served. 
* It feels less invasive. If the package would ship a bundle itself, it would look also like this. 

Con:
* The physical paths are not correct for the worker and asset plugin. They (and other plugins) need to resolve the source of the dependency bundle if the URL points to a physical file (this PR).

_/playground/assets/node_modules/.vite/deps/@vitejs_test-dep-with-asset.js?v=1cdbd667_
```js
// ../../node_modules/.pnpm/@vitejs+test-dep-with-asset@file+playground+assets+dep-with-asset/node_modules/@vitejs/test-dep-with-asset/dir/path.js
var imageUrl = new URL("./asset.png", import.meta.url).href;

// ../../node_modules/.pnpm/@vitejs+test-dep-with-asset@file+playground+assets+dep-with-asset/node_modules/@vitejs/test-dep-with-asset/index.js
var imageUrlRoot = new URL("./asset.png", import.meta.url).href;
export {
  imageUrl,
  imageUrlRoot
};
//# sourceMappingURL=@vitejs_test-dep-with-asset.js.map
```

**After Bundling (my understanding of your approach)**

The URLs are adjusted to relative (or absolute?) file paths on disk. 

Pro: 
* URLs will point to the correct file paths after bundling (relative to the bundle) allowing easy resolving of paths in worker and asset plugins. 

Con:
* Another plugins MUST pickup these URLs and ensure they are further transformed and served. Otherwise URLs might attempt to reach beyond the root path pointing to a path which might not be served `new URL('../../../../../node_modules/@depscope/dep/subpath/test.html', import.meta.url).href` where `import.meta.url=http://localhost/app.js`
* If 3rd party plugins want to do a custom URL logic not based on file paths they have to "undo" the bundler path logic. 

Code: 
_/playground/assets/node_modules/.vite/deps/@vitejs_test-dep-with-asset.js?v=1cdbd667_
```js
// ../../node_modules/.pnpm/@vitejs+test-dep-with-asset@file+playground+assets+dep-with-asset/node_modules/@vitejs/test-dep-with-asset/dir/path.js
var imageUrl = new URL("../../../../../node_modules/.pnpm/@vitejs+test-dep-with-asset@file+playground+assets+dep-with-asset/node_modules/@vitejs/test-dep-with-asset/asset.png", import.meta.url).href;

// ../../node_modules/.pnpm/@vitejs+test-dep-with-asset@file+playground+assets+dep-with-asset/node_modules/@vitejs/test-dep-with-asset/index.js
var imageUrlRoot = new Url("../../../../../node_modules/.pnpm/@vitejs+test-dep-with-asset@file+playground+assets+dep-with-asset/node_modules/@vitejs/test-dep-with-asset/asset.png", import.meta.url).href;
export {
  imageUrl,
  imageUrlRoot
};
//# sourceMappingURL=@vitejs_test-dep-with-asset.js.map
```

For my own needs I am fine with either approach. For my needs even only this PR is enough. 

I am only a bit worried about unexpected side effects of the bundler rewriting too much requiring us to revert the change again. A configuration option (magic comments / vite.config with regexp patterns) to control the behavior could reduce the risk at cost of complexity in Vite delivering correct results in all combinations. 

As owners of Vite its your call for decision. I'd be happy to give either one a try to implement.

Reply:
We discussed this PR in today's team meeting.

We think this issue is better solved at esbuild level (or Rolldown level after we start using it) and would benefit more people. So we suggest helping esbuild supporting this feature (https://github.com/evanw/esbuild/pull/2508). For example, providing a test case.

But if you would like to work on the fix on Vite side, we are open to having a fix on Vite side in the way 2 or 3 in https://github.com/vitejs/vite/pull/16418#discussion_r1661124403. It was mentioned in the meeting that [`onEnd`](https://esbuild.github.io/plugins/#on-end) might also be useful. But if the implementation is complex, we might not merge the PR. Even if the PR is merged, we would partially or fully revert the PR in the near future when we start using Rolldown in Vite for the optimizer.